pkg/sloop/ingress/kubewatcher.go

```diff
@@ -81,77 +100,149 @@ func NewKubeWatcherSource(kubeClient kubernetes.Interface, outChan chan typed.Ku
+
+	crdInformers := i.crdInformers
+	i.crdInformers = make(map[crdGroupVersionResourceKind]*crdInformerInfo)
+	return crdInformers
+}
 
-		informer.Informer().AddEventHandler(i.getEventHandlerForResource(resourceKind))
+func (i *kubeWatcherImpl) existingOrStartNewCrdInformer(crd crdGroupVersionResourceKind, existing map[crdGroupVersionResourceKind]*crdInformerInfo, factory dynamicinformer.DynamicSharedInformerFactory) {
+	i.protection.Lock()
+	defer i.protection.Unlock()
+	if i.stopped {
+		return
+	}
 
-		go func() {
-			glog.V(2).Infof("Starting CRD informer for: %s (%v)", resourceKind, resource)
-			informer.Informer().Run(i.stopChan)
-			glog.V(2).Infof("Exited CRD informer for: %s", resourceKind)
-		}()
+	// if there is an existing informer for this crd, then keep using the existing informer
+	crdInformer, found := existing[crd]
+	if found {
+		i.crdInformers[crd] = crdInformer
+		delete(existing, crd) // remove from existing so it wont get stopped as unwanted
+		return
 	}
 
-	return nil
+	// need an informer for this crd
+	crdInformer = &crdInformerInfo{crd: crd, stopChan: make(chan struct{})}
+	i.crdInformers[crd] = crdInformer
+	i.startNewCrdInformer(crdInformer, factory)
+}
+
+func (i *kubeWatcherImpl) startNewCrdInformer(crdInformer *crdInformerInfo, factory dynamicinformer.DynamicSharedInformerFactory) {
+	gvr := schema.GroupVersionResource{Group: crdInformer.crd.group, Version: crdInformer.crd.version, Resource: crdInformer.crd.resource}
+	kind := crdInformer.crd.kind
+	informer := factory.ForResource(gvr)
+	informer.Informer().AddEventHandler(i.getEventHandlerForResource(kind))
+
+	go func() {
+		glog.V(2).Infof("Starting CRD informer for: %s (%v)", kind, gvr)
+		metricCrdInformerRunning.Set(float64(atomic.AddInt64(&i.activeCrdInformer, 1)))
+
+		informer.Informer().Run(crdInformer.stopChan)
+
+		glog.V(2).Infof("Exited CRD informer for: %s (%v)", kind, gvr)
+		metricCrdInformerRunning.Set(float64(atomic.AddInt64(&i.activeCrdInformer, -1)))
+	}()
 }
 
-func getCrdList(kubeCfg *rest.Config) ([]crdGroupVersionResourceKind, error) {
-	crdClient, err := clientset.NewForConfig(kubeCfg)
+func stopUnwantedCrdInformers(existing map[crdGroupVersionResourceKind]*crdInformerInfo) {
+	// no lock is needed - all these informers should be disconnected from kubeWatcherImpl
+	for _, v := range existing {
+		gvr := schema.GroupVersionResource{Group: v.crd.group, Version: v.crd.version, Resource: v.crd.resource}
+		glog.V(2).Infof("Stopping CRD informer for: %s (%v)", v.crd.kind, gvr)
+		close(v.stopChan)
+	}
+}
+
+func getCrdList(crdClient clientset.Interface) ([]crdGroupVersionResourceKind, error) {
+	crdList, err := crdClient.ApiextensionsV1().CustomResourceDefinitions().List(metav1.ListOptions{})
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to instantiate client for querying CRDs")
+		glog.Errorf("Failed to get CRD list from ApiextensionsV1, falling back to ApiextensionsV1beta1")
+		return getCrdListV1beta1(crdClient)
```

Comment:
What's the thought behind keeping 2 functionalities to get crd list ? safety net ?

Reply:
the schema changed - `getCrdListV1beta1` used the previous schema, this one is current.
If the environment that this code runs in is old, this method will fail and the other one should work